src/NHibernate/Driver/DriverBase.cs

```diff
@@ -231,7 +231,7 @@ public void RemoveUnusedCommandParameters(DbCommand cmd, SqlString sqlString)
 				.ForEach(unusedParameterName => cmd.Parameters.RemoveAt(unusedParameterName));
 		}
 
-		public virtual void ExpandQueryParameters(DbCommand cmd, SqlString sqlString)
+		public virtual void ExpandQueryParameters(DbCommand cmd, SqlString sqlString, SqlType[] parameterTypes)
```

Comment:
This method in the case of ODBC was replacing the original parameters by new one, partially clone, leaving aside `Scale`, `Precision` and `Size` set by `OdbcDriver.InitializeParamater`. So I wonder if those changes in `InitializeParamater` were tested. It does not seem to be the case. When looking at [NH-3895](https://nhibernate.jira.com/browse/NH-3895) which supposed to be fixed by one of those changes, there are no test in the commit.

Due to `ExpandQueryParameters`, `Scale` is lost. Due to the change in `DateTimeType`, we now have milliseconds, and this has triggered the NH-3895 bug for numerous tests. So I have changed `ExpandQueryParameters` for it to create parameters the same way as original one, calling `InitializeParamater`. This fixes NH-3895. But it has triggered the trouble explained in previous comment.

Reply:
Having dug into files history, the size setting has been added with 03eb58acae867e2d5620446c3b07cb91e7d74422 in January 2010, while the `ExpandQueryParameters` was not yet there. This one has been added in July 2010 with 131b97e9971608e5f90b31cd57e09b8d7935c4a8. So, not having Size/Precision/Scale actually specified for Odbc parameters (and all other drivers not handling named parameters - DB2, Ifx, Ingres, OleDb, OracelLite, SybaseAsa - if they do set them in `InitializeParameter`,) is an old regression.

Reply:
But later on [NH-3036](https://nhibernate.jira.com/browse/NH-3036) reverted the size setting for sql client due to issues with `like` (which may requires larger size than the actual column size), so I think we should remove this `Size` setting from Odbc parameters too.